### DISCUSSION TITLE
Removing "paper" noise from scan

### DISCUSSION THREAD

**Discussion by xanatos** [ImageMagick/ImageMagick#6659] [Digital Image Processing]
I have a scan. I want to remove what I think is "paper" noise:

![image](https://github.com/ImageMagick/ImageMagick/assets/821344/b4a53c6b-bff2-400a-bfb9-6f31880696cb)

How can I do? Thanks!

**Comment by fmw42** [comment]
This dither pattern can be **mitigated** (not removed completely) using FFT notch filtering. Since the dither pattern is regular and uniform across the image, it should be apparent as bright spots in the FFT spectrum (log of magnitude) image. These spots can then be masked out in the magnitude after creating a mask that has the spots blocked out.  Here are the individual steps and resulting images.

\# Get image dimensions
```
declare `magick image.png -format "W=%w\nH=%h\n" info:`
echo "W=$W; H=$H;"
```

\# Get magnitude and phase images from -fft
`magick image.png -alpha off -fft mag-phase.png `

\# Get spectrum image (log of magnitude)
`magick mag-phase-0.png -evaluate log 10000 temp1.png`

![temp1](https://github.com/ImageMagick/ImageMagick/assets/13458218/ba0c39ea-03fe-473d-82df-8b5058739d74)

\# Dilate the spectrum to make bright dots into bright circles
`magick temp1.png -morphology dilate disk:5 temp2.png`

![temp2](https://github.com/ImageMagick/ImageMagick/assets/13458218/48acd943-805b-40a1-90ec-109ab13186c1)

\# Get center coordinates for spectrum images 
\# (Note spectrum will be made square from largest dimension of input)
```
cent=`magick xc: -format "%[fx:max($W,$H)/2]" info:`
echo "cent=$cent;"
```

\# Draw a large black circle around the low frequency region, but not covering the dither pattern bright spots
`magick temp2.png -fill black -draw "translate $cent,$cent circle 0,0 0,55" -alpha off temp3.png`

![temp3](https://github.com/ImageMagick/ImageMagick/assets/13458218/e7874dd4-bd1b-4474-aa9f-e9f393de6730)


\# Threshold and negate to keep only the circles from the dither pattern as dark circles on white background as a mask
 `magick temp3.png -threshold 25% -negate mask.png`

![mask](https://github.com/ImageMagick/ImageMagick/assets/13458218/c6a8bdef-6492-436c-a09a-a9b963eae198)


\# Apply the mask to the original magnitude and combine with the original phase and do the IFT and crop to original input dimensions to get the final result
`magick \( mag-phase-0.png mask.png -compose multiply -composite \) mag-phase-1.png -ift -crop ${W}x${H}+0+0 +repage result.png`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/fd203c8e-7bab-4226-b575-1560907be589)


\#Now we can combine this all into two commands as follows:
```
declare `magick image.png -format "W=%w\nH=%h\nCENT=%[fx:max($W,$H)/2]" info:`
echo "W=$W; H=$H; CENT=$CENT;"

magick image.png -alpha off -fft -write mpr:mag-phase -delete 0,1 \
\( mpr:mag-phase[0] -evaluate log 10000 \
-morphology dilate disk:5 \
-fill black -draw "translate $CENT,$CENT circle 0,0 0,55" -alpha off \
-threshold 25% -negate -write mpr:mask +delete \) \
\( mpr:mag-phase[0] mpr:mask -compose multiply -composite \) \
mpr:mag-phase[1] -ift -crop ${W}x${H}+0+0 +repage \
result.png

```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/0ec04e19-e93e-4ddb-80f0-ab26f7748b44)


**IMPORTANT** 
Note that -fft/-ift produces erroneous results in IM 7.1.1.16 and 17 (and corresponding IM 6 versions). The above was produced with IM 7.1.1.15.

**Comment by xanatos** [comment]
Sadly I still have to try your solution. I don't have a Linux machine with enough memory for my image, and the Windows version of IM doesn't have the fftw library. I'm trying to recompile IM (an older version, because as you noticed there is a bug in fft), but I'm having problems.

**Reply by fmw42** [reply]
A beta is supposed to be available tomorrow with the patch.

**Reply by fmw42** [reply]
You can also test with a smaller image that would work on the Linux machine that you have.

**Reply by fmw42** [reply]
Some argument may need adjusting for different image patterns and sizes.

**Reply by xanatos** [reply]
Finally was able to recompile for Windows with fft. Tested with the image I had posted and your commands (I have a bash for Windows), and the results were the same as yours. But for bigger pieces of the image the results get worse and worse:

From:

![image](https://github.com/ImageMagick/ImageMagick/assets/821344/ff37f64e-6872-4b37-81eb-ddaad4793ade)

I get:

![result](https://github.com/ImageMagick/ImageMagick/assets/821344/a493b882-5b47-428b-82bc-b3f8d2ebce52)

**Reply by fmw42** [reply]
It what regard is this worse.  It has mitigated (reduced the effect) of the dither pattern.

What was your command?  Was it exactly the same as mine?  Perhaps you need to adjust some arguments.  That would need to be done with the first set of commands that I provided so that you can see the effects of each step.

**Comment by fmw42** [comment]
I adjusted two of the arguments (radius and threshold) to tweak the result.

```
declare magick image.png -format "W=%w\nH=%h\nCENT=%[fx:max($W,$H)/2]" info:`
echo "W=$W; H=$H; CENT=$CENT;"

magick image.png -alpha off -fft -write mpr:mag-phase -delete 0,1 \
\( mpr:mag-phase[0] -evaluate log 10000 \
-morphology dilate disk:5 \
-fill black -draw "translate $cent,$cent circle 0,0 0,75" -alpha off \
-threshold 20% -negate +write mask.png -write mpr:mask +delete \) \
\( mpr:mag-phase[0] mpr:mask -compose multiply -composite \) \
mpr:mag-phase[1] -ift -crop ${W}x${H}+0+0 +repage \
result.png

```


Mask:

![mask](https://github.com/ImageMagick/ImageMagick/assets/13458218/579e639e-3428-42fb-abfb-5bb28db8f6a5)


Result:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/1a6d8f60-609e-42df-8d69-3a3742740963)

**Comment by fmw42** [comment]
Here is another attempt.  I have masked out the center columns and center rows in the code so the blacked out image looks as follows, if you process one line at a time. So it just has the circle (dilated) spots representing the dither pattern in the spectrum

![temp3](https://github.com/ImageMagick/ImageMagick/assets/13458218/a7833066-2594-4576-85fd-f61aa16aab5b)


Here is the two line code:

```
declare `im71115 magick image.png -format "W=%w\nH=%h\nCENT=%[fx:max($W,$H)/2]" info:`
echo "W=$W; H=$H; CENT=$CENT;"

im71115 magick image.png -alpha off -fft -write mpr:mag-phase -delete 0,1 \
\( mpr:mag-phase[0] -evaluate log 10000 \
-morphology dilate disk:5 \
-fill black -draw "translate $CENT,$CENT circle 0,0 0,75" \
-draw "rectangle 0,%[fx:$CENT-5] %w,%[fx:$CENT+5] rectangle %[fx:$CENT-5],0 %[fx:$CENT+5],%h" -alpha off \
-threshold 20% -negate +write mask.png -write mpr:mask +delete \) \
\( mpr:mag-phase[0] mpr:mask -compose multiply -composite \) \
mpr:mag-phase[1] -ift -crop ${W}x${H}+0+0 +repage \
result.png

```

Result:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/65496512-79b4-4df6-9dbd-bc6620b73cef)

**Reply by xanatos** [reply]
Thanks!

**Comment by fmw42** [comment]
The issue with -fft -ift seems to be fixed in IM 7.1.1.8
